File: Excel/456 Challenge.R
library(tidyverse)
library(readxl)

input = read_excel("Excel/456 Extract special Characters.xlsx", range = "A1:A10")
test  = read_excel("Excel/456 Extract special Characters.xlsx", range = "B1:B10")

# approach 1 - remove alphanumerics 

result = input %>%
  mutate(String = str_replace_all(String, "[[:alnum:]]", "")) %>%
  mutate(String = ifelse(String == "", NA, String))

identical(result$String, test$`Expected Answer`)
#> [1] TRUE

# approach 2 - extract special characters

result2 = input %>%
  mutate(String = str_extract_all(String, "[^[:alnum:]]") %>%
           map_chr(~paste(.x, collapse = ""))) %>%
  mutate(String = ifelse(String == "", NA, String))

identical(result2$String, test$`Expected Answer`)
#> [1] TRUE

File: Excel/456 Challenge.py
import pandas as pd
import numpy as np

input = pd.read_excel('456 Extract special Characters.xlsx', usecols="A", nrows = 10)
test  = pd.read_excel('456 Extract special Characters.xlsx', usecols="B", nrows = 10)

# Approach 1 - remove alphanumeric characters
result = input['String'].str.replace(r'[a-zA-Z0-9]', '', regex=True)
result = result.replace(r'^\s*$', np.nan, regex=True)

print(result.equals(test['Expected Answer'])) # True

# Approach 2 - extract special characters
result2 = input['String'].str.extractall(r'([^a-zA-Z0-9])')[0].unstack()
result2 = result.replace(r'^\s*$', np.nan, regex=True)

print(result2.equals(test['Expected Answer'])) # True
